### PR TITLE
SOMA-level dimension slicing

### PR DESCRIPTION
**Pull request by aaronwolen**
This adds a new `set_query()` method to define the ranges of the indexed dimensions to slice to:

- `TileDBArray` and its subclasses
- `AnnotationGroup` and its subclasses
- `SOMA`

See the new [*Filtering* vignette](https://github.com/TileDB-Inc/tiledbsc/blob/aaronwolen/sc-18255/implement-soma-level-dimension-slicing/vignettes/filtering.Rmd) for details and examples.

Other changes:

- `AnnotationDataframe` class now has `ids()` method to retrieve all dimension values
- Added `set_query` method to `TileDBArray`, `AnnotationGroup`, and `SOMA`
- Added tests for basic mechanics of array dim slicing
- Added missing config/ctx args to AnnotationGroup
- Internal test helper `create_test_group_with_members()` is now more configurable and creates arrays with string dimensions
- Added a utility to assert subsets and provide informative errors

### CODE REVIEW

**Comment by aaronwolen** [comment]
CC @dnadave @kaitlin-procogia @augustine-procogia

File: vignettes/filtering.Rmd [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+designed to work with the `pbmc_small` dataset.
+
+To see this in action, let's create a `SOMA` from `pbmc_small`'s `RNA` assay.
+
+```{r}
+soma <- SOMA$new(uri = file.path(tempdir(), "pbmc_small_rna"))
+soma$from_seurat_assay(pbmc_small[["RNA"]])
+```
+
+Now let's use `set_query()` to slice 3 features of interest.
```
Comment (johnkerl) [inline vignettes/filtering.Rmd:79]: 3 -> three

File: vignettes/filtering.Rmd [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+```{r}
+soma$set_query(var_ids = c("PPBP", "PF4", "GNLY"))
+```
+
+Any of the individual SOMA components will respect the new query. So if we ready in the feature-level metadata from `var`, the resulting data frame contains only the features we specified.
```
Comment (johnkerl) [inline vignettes/filtering.Rmd:85]: ready -> read

File: vignettes/filtering.Rmd [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+```
+
+And when we recreate the entire Seurat `Assay` from the `SOMA`, the slice is carried through to all relevant components,
+
+```{r}
+pbmc_small_rna <- soma$to_seurat_assay()
+```
+
+including the raw counts,
+
```
Comment (johnkerl) [inline vignettes/filtering.Rmd:98]: extra blank line

**Review by johnkerl — APPROVED** [review]
LGTM -- nits is all I found :D

File: R/AnnotationDataframe.R [hunk 4]
```diff
@@ -13,6 +13,14 @@ AnnotationDataframe <- R6::R6Class(
     #' @field verbose Print status messages
     verbose = TRUE,
 
+    #' @description Retrieves the values from the array's dimension.
+    ids = function() {
+      arr <- self$object
+      # TODO: Workaround bug using tiledb::attrs(arr) <- NA_character_
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply R/AnnotationDataframe.R:19]: Yeap, noted in the comments.

File: tests/testthat/test-SOMA_dim_slice.R [hunk 5]
```diff
@@ -0,0 +1,40 @@
+test_that("SOMA object can be sliced by dimension", {
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply tests/testthat/test-SOMA_dim_slice.R:1]: Yeah the inconsistency is gross. I started with `test_` but testthat's convention is `test-`, so I'm slowly migrating to that.

File: vignettes/filtering.Rmd [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+We can use the `set_query()` method to read in a specific subset of cells by providing a named list of ranges for each dimension.
+
+```{r}
+adf$set_query(
+  dims = list(obs_id = c("AAATTCGAATCACG", "AAGCAAGAGCTTAG", "AAGCGACTTTGACG"))
+)
+```
+
+Now, the next time we call `to_dataframe()` the resulting data frame will only contain the cells we specified.
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply vignettes/filtering.Rmd:58]: I don't have a strong opinion but its "data frame" in base R's docs.

**Pull request merged**